### Apriori Algorithm Example and Logic

The Apriori algorithm can be broken down into a basic If ----> Then concept. IF a customer purchases item A THEN what is the liklihood they also purchase item B? The 'THEN' condition is broken down as a number as in, IF customer purchases bread THEN 60% chance they buy jam/jelly. Thats the jist of what we are trying to accomplish. The "IF" is referred to as the "antecedent(s)" and the "THEN", "Consequent(s)".

**Imporant to remember we are not looking for a causality, but rather a pattern of co-occurence**

There are 3 matrices that help to measure the association between antecedent(s) and consequent(s):
- Support
- Condifence
- Lift

**Support** is the frequency of item 1 or the combination of item1 and item2 : Its basically the frequency of the items which we have bought and what are the combination(s) of the frequency of the item we have bought.

**Confidence** tells us how often the items 1 & 2 occur together given the number of times item 1 occurs.

**Lift** is essentially the strength of any rule

In [2]:
import pandas as pd

In [3]:
data = [['T1', 'A', 'B','C'], ['T2', 'A', 'C', 'D'], ['T3', 'B', 'C', 'D'], ['T4', 'A', 'D', 'E'], ['T5', 'B', 'C', 'E']]
example_data = pd.DataFrame(data)
example_data

0  1  2  3
0  T1  A  B  C
1  T2  A  C  D
2  T3  B  C  D
3  T4  A  D  E
4  T5  B  C  E

## Theortical Rules

In [4]:
cols1 = ['Values', 'Frequency/Support']
data3 = [['A', 3], ['B',3], ['C',4], ['D',3], ['E',2]]
value_support_table = pd.DataFrame(data3, columns = cols1)
value_support_table

Values  Frequency/Support
0      A                  3
1      B                  3
2      C                  4
3      D                  3
4      E                  2

In [5]:
cols = ['Rule', 'Support', 'Confidence', 'Lift', 'Supp_Num', 'Conf_Num', 'Lift_Num']
data1 = [['A==>D', '2/5', '2/3', '10/9', (2/5), (2/3), (10/9)], ['C==>A', '2/5', '2/4', '5/6', (2/5), (2/4), (5/6)], ['A==>C', '2/5', '2/3', '5/6',  (2/5), (2/3), (5/6)], ['B&C==>A', '1/5', '1/3', '5/9',  (1/5), (1/3), (5/9)]] 
rules_explanation_table = pd.DataFrame(data1, columns = cols)
rules_explanation_table

Rule Support Confidence  Lift  Supp_Num  Conf_Num  Lift_Num
0    A==>D     2/5        2/3  10/9       0.4  0.666667  1.111111
1    C==>A     2/5        2/4   5/6       0.4  0.500000  0.833333
2    A==>C     2/5        2/3   5/6       0.4  0.666667  0.833333
3  B&C==>A     1/5        1/3   5/9       0.2  0.333333  0.555556

### Row 0 Analysis

-Row 0 Support ---> A & D appear together 2 times in all of our 5 transactions ---> 40% of the time

-Row 0 Confidence ----> A & D appear together 2 times and A appears in 3 total transactions ---> 66% of the time

-Row 0 Lift ----> (Rule Confidence/Benchmark Confidence)
   - Benchmark Confidence = (Support of Consequent(A=3)/Total_Number_Transactions(5); 3/5 = .6
   - Rule_Confidence(2/3 = 0.666667)/Benchmark_Confidence(.6)
   - Lift = (0.666667)/(0.6) =====> 1.1111111(Lift_num)
    
### Thus....
The rule of if A then D has a support of .4 which means in the COMBINATION of A and D appears in 40% of ALL of the transactions.

The rule of if A then D has a confidence value of 0.666667 which means that the COMBINATION of A and D appears in 67% of the transactions in which A was purchased.

The rule of if A then D has a lift value of 1.1111 which means that that there is a strong association for this rule. The lift value expresses that the presence of D give that A is present happens more often than would be expected. 
- (Confidence / (Expected/Benchmark Confidence)